ballista/rust/executor/src/cpu_bound_executor.rs

```diff
@@ -0,0 +1,376 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//Inspire by https://thenewstack.io/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/
+
+//! This module contains a dedicated thread pool for running "cpu
+//! intensive" workloads as query plans
+
+use log::warn;
+use parking_lot::Mutex;
+use std::{pin::Pin, sync::Arc};
+use tokio::sync::oneshot::Receiver;
+
+use futures::Future;
+
+/// The type of thing that the dedicated executor runs
+type Task = Pin<Box<dyn Future<Output = ()> + Send>>;
+
+/// Runs futures (and any `tasks` that are `tokio::task::spawned` by
+/// them) on a separate tokio runtime, like separate CPU-bound (execute a datafusion plan) tasks
+/// from IO-bound tasks(heartbeats). Get more from the above blog.
+#[derive(Clone)]
+pub struct DedicatedExecutor {
+    state: Arc<Mutex<State>>,
+}
+
+/// Runs futures (and any `tasks` that are `tokio::task::spawned` by
+/// them) on a separate tokio Executor
+struct State {
+    /// The number of threads in this pool
+    num_threads: usize,
+
+    /// The name of the threads for this executor
+    thread_name: String,
+
+    /// Channel for requests -- the dedicated executor takes requests
+    /// from here and runs them.
+    requests: Option<std::sync::mpsc::Sender<Task>>,
+
+    /// The thread that is doing the work
+    thread: Option<std::thread::JoinHandle<()>>,
+}
+
+/// The default worker priority (value passed to `libc::setpriority`);
+const WORKER_PRIORITY: i32 = 10;
+
+impl std::fmt::Debug for DedicatedExecutor {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let state = self.state.lock();
+
+        let mut d = f.debug_struct("DedicatedExecutor");
+
+        d.field("num_threads", &state.num_threads)
+            .field("thread_name", &state.thread_name);
+
+        if state.requests.is_some() {
+            d.field("requests", &"Some(...)")
+        } else {
+            d.field("requests", &"None")
+        };
+
+        if state.thread.is_some() {
+            d.field("thread", &"Some(...)")
+        } else {
+            d.field("thread", &"None")
+        };
+
+        d.finish()
+    }
+}
+
+impl DedicatedExecutor {
+    /// https://stackoverflow.com/questions/62536566
+    /// Creates a new `DedicatedExecutor` with a dedicated tokio
+    /// runtime that is separate from the `[tokio::main]` threadpool.
+    ///
+    /// The worker thread priority is set to low so that such tasks do
+    /// not starve other more important tasks (such as answering health checks)
+    ///
+    pub fn new(thread_name: impl Into<String>, num_threads: usize) -> Self {
```

Comment:
FWIW I think @Darksonn and others have noted that a tokio `Handle` might also be able to be used here https://docs.rs/tokio/1.17.0/tokio/runtime/struct.Handle.html